ADAL/src/ADAuthorityValidation.m

```diff
@@ -0,0 +1,396 @@
+    
+    _validatedADAuthorities = [NSMutableSet new];
+    //List of prevalidated authorities (Azure Active Directory cloud instances).
+    //Only the sThrustedAuthority is used for validation of new authorities.
+    [_validatedADAuthorities addObject:[NSURL URLWithString:s_kTrustedAuthority]];
+    [_validatedADAuthorities addObject:[NSURL URLWithString:s_kTrustedAuthorityChina]]; // Microsoft Azure China
+    [_validatedADAuthorities addObject:[NSURL URLWithString:s_kTrustedAuthorityGermany]]; // Microsoft Azure Germany
+    [_validatedADAuthorities addObject:[NSURL URLWithString:s_kTrustedAuthorityWorldWide]]; // Microsoft Azure Worldwide
+    [_validatedADAuthorities addObject:[NSURL URLWithString:s_kTrustedAuthorityUSGovernment]]; // Microsoft Azure US Government
+    
+    return self;
+}
+
+- (void)dealloc
+{
+    SAFE_ARC_RELEASE(_validatedADAuthorities);
+    _validatedADAuthorities = nil;
+    
+    SAFE_ARC_RELEASE(_validatedAdfsAuthorities);
+    _validatedAdfsAuthorities = nil;
+    
+    SAFE_ARC_SUPER_DEALLOC();
+}
+
+
+#pragma mark - caching
+- (BOOL)addValidAuthority:(NSURL *)authority domain:(NSString *)domain
+{
+    if (!domain || !authority)
+    {
+        return NO;
+    }
+    
+    // Get authorities for domain (UPN suffix) and create one if needed
+    NSMutableSet *authorities = [_validatedAdfsAuthorities objectForKey:domain];
+    if (!authorities)
+    {
+        authorities = [NSMutableSet new];
+        [_validatedAdfsAuthorities setObject:authorities forKey:domain];
+    }
+  
+    // Add given authority to trusted set for the domain
+    [authorities addObject:authority];
+    return YES;
+}
+
+- (BOOL)addValidAuthority:(NSURL *)authority
+{
+    if (!authority)
+    {
+        return NO;
+    }
+    [_validatedADAuthorities addObject:authority];
+    return YES;
+}
+
+
+- (BOOL)isAuthorityValidated:(NSURL *)authority domain:(NSString *)domain
+{
+    // Check for authority
+    NSSet *authorities = [_validatedAdfsAuthorities objectForKey:domain];
+    for (NSURL *url in authorities)
+    {
+        if([url isEquivalentAuthority:authority])
+        {
+            return YES;
+        }
+    }
+    return NO;
+}
+
+// Checks the cache for previously validated authority.
+// Note that the authority host should be normalized: no ending "/" and lowercase.
+- (BOOL)isAuthorityValidated:(NSURL *)authority
+{
+    if (!authority)
+    {
+        return NO;
+    }
+    for (NSURL *url in _validatedADAuthorities)
+    {
+        if([url isEquivalentAuthority:authority])
+        {
+            return YES;
+        }
+    }
+    return NO;
+}
+
+
+#pragma mark - Authority validation
+
+- (void)validateAuthority:(NSString *)authority
+            requestParams:(ADRequestParameters *)requestParams
+          completionBlock:(void (^)(BOOL validated, ADAuthenticationError *error))completionBlock
+{
+    _correlationId = requestParams.correlationId;
+    _telemetryRequestId = requestParams.telemetryRequestId;
```

Comment:
Making these properties on singleton objects makes me nervous, it's better to pass around `id<ADRequestContext>` in that case.